### PR TITLE
ISSUE-19 mining classification

### PR DESCRIPTION
**Pull request by ghost**
this is a work-in-progress pull request for [mining classification](https://github.com/capstone-coal/pycoal/issues/19). it is based off of my unmerged branch for [environmental correlation](https://github.com/capstone-coal/pycoal/pull/24), so you can focus on `mining.py` and ignore the other files.

i [shared](https://github.com/capstone-coal/pycoal/issues/19#issuecomment-277969720) this skeleton with @xiaomei7 on 2017-02-07, and i am creating this pull request because there have been no updates. i will take responsibility for this module if she does not, otherwise i will gladly cede this pull request to her and collaborate as needed.

note that there are interdependencies with @heidiaclayton's efforts at [mineral classification](https://github.com/capstone-coal/pycoal/pull/23) and my implementation of GIS processing, making it necessary to coordinate development.

assigning to myself and requesting review from @lewismc . i am aware that the project structure and class hierarchy will need to be updated as mentioned in the other issues.

### CODE REVIEW

**Comment by lewismc** [comment]
@browtayl please update the conflicts. Thank you

**Comment by ghost** [comment]
@lewismc will do.

my goal is to make significant progress on this and the [other](https://github.com/capstone-coal/pycoal/pull/24) pull request by thursday.

**Comment by ghost** [comment]
fixed conflicts by merging `browtayl:coal-20-environmental-correlation`. again, you can ignore everything except `mining.py`.

when it comes time to merge into `master` we may want to merge that branch first. i probably should have implemented the GIS stuff in a separate branch so we don't have this interdependency. i can still do that if desired.

**Comment by ghost** [comment]
this pull request now contains the most up-to-date code for `mining.py`, `gis.py`, and `environment.py`. i fixed the conflicts and added classes. the code is still skeletal, but currently those files are empty in `master`.

@lewismc request review and approval to merge. this will also close https://github.com/capstone-coal/pycoal/pull/24. i think i will be able to implement additional functionality more cleanly in a separate branch.

File: pycoal/environment.py
Comment (lewismc) [inline pycoal/environment.py]: Please add ALv2.0 header

File: pycoal/environment.py
Comment (lewismc) [inline pycoal/environment.py]: If possible, then please implement documentation so we have a better opportunity to discuss this class on Friday. If possible, please do this for all functions.

File: pycoal/environment.py
Comment (lewismc) [inline pycoal/environment.py:24]: Is it just me or has the formatting/indent changed here from 4 space/tab to 8 space/2 tab?

File: pycoal/gis.py
Comment (lewismc) [inline pycoal/gis.py]: License header please.

File: pycoal/mining.py
Comment (lewismc) [inline pycoal/mining.py]: License header please.

**Review by lewismc — CHANGES_REQUESTED** [review]
This is looking good. Please just add license headers and if possible also add the gdal dependency to setup.py.
After that I am happy for you to merge into master and we can walk through the unknowns on Friday. Excellent work.

File: pycoal/environment.py
Comment (ghost) [reply pycoal/environment.py:24]: @lewismc i indented the members of the class, so everything should be shifted. there might be an error though.

File: pycoal/environment.py
Comment (ghost) [reply pycoal/environment.py]: @lewismc will do

File: pycoal/environment.py
Comment (ghost) [reply pycoal/environment.py]: @lewismc forgot about that, sorry. will add before merge.

Comment (ghost) [reply pycoal/environment.py]: @lewismc i notice the other source files aren't using `#!/usr/bin/env python`. do we need this, or should i remove it?

**Comment by ghost** [comment]
@lewismc added license headers

**Comment by ghost** [comment]
@lewismc added gdal dep

**Comment by ghost** [comment]
build is failing due to invalid syntax

edit: or maybe it's due to the gdal installation...

**Comment by ghost** [comment]
@lewismc one thing we'll have to watch out for with the gdal dep is the correct version. right now the rotation bug is only fixed in `trunk`.

**Comment by ghost** [comment]
sorry that these files are still sort of a mess. they don't reflect a lot of the things i learned since i first wrote them and will need to be heavily revised. the environmental correlation in particular may have to be done differently since the data points are [sparse](https://groups.google.com/forum/#!topic/coal-capstone/MBVIBZBc_cs). i am soft-blocked by the lack of mineral classified data.

**Comment by lewismc** [comment]
@browtayl no problem about the file right now, we will work to get them substantiated. I'll check out the build. Feel free the merge and I am happy for master to be broken until we can populate the unknowns.

File: pycoal/environment.py
Comment (lewismc) [reply pycoal/environment.py]: No wen don't need it.

Comment (ghost) [reply pycoal/environment.py]: @lewismc removed the crunchbangs

**Comment by lewismc** [comment]
Builds are currently failing due to the ```???```'s which are present.

**Comment by ghost** [comment]
so i addressed all the points except adding more detailed docstrings which will probably need to be changed. i will merge this and then work on making it build (and run correctly) as soon as possible.

**Comment by ghost** [comment]
added fix to change the imports from `coal` to `pycoal` and use the new class names.

**Comment by ghost** [comment]
@lewismc merging and breaking `master` as approved.

this excerpt from *Test Driven Development By Example* seems apropos:
> Broken Test
> 
> How do you leave a programming session when you're programming alone? Leave the last test broken.
> 
> Richard Gabriel taught me the trick of finishing a writing session in midsentence. When you sit back down, you look at the half-sentence and you have to figure out what you were thinking when you wrote it. Once you have the thought thread back, you finish the sentence and continue. Without the urge to finish the sentence, you can spend many minutes first sniffing around for what to work on next, then trying to remember your mental state, then finally getting back to typing.
> 
> I tried the analogous technique for my solo projects, and I really like the effect. Finish a solo session by writing a test case and running it to be sure it doesn't pass. When you come back to the code, you then have an obvious place to start. You have an obvious, concrete bookmark to help you remember what you were thinking; and making that test work should be quick work, so you'll quickly get your feet back on that victory road.
> 
> I thought it would bother me to have a test broken overnight. It doesn't, I think because I know that the program isn't finished. A broken test doesn't make the program any less finished, it just makes the status of the program manifest. The ability to pick up a thread of development quickly after weeks of hiatus is worth that little twinge of walking away from a red bar.

this merge is likewise a **half-complete thought**, and there are many points yet to elaborate in addition to obvious errors. in particular:
* we will have to **rethink environmental correlation** if we cannot assume we have a nice direct overlay. if we use the [water quality data](https://groups.google.com/forum/#!topic/coal-capstone/MBVIBZBc_cs) there are opportunities for automation such as hydrographic gradients, but i'm not sure how valuable that would be beyond our case study.
* the **GIS** code is a good starting point, but i will need to revisit some of the assumptions. one thing i want to do is remove the `selectLayers` parameter to `combine`. this method will be useful, after some modification, for **generating training data** for mining id: combine a mine boundary polygon with a classified raster and train. i think i may end up cropping the other way though, using the vector rather than the raster as the boundary. as i mentioned [before](https://groups.google.com/d/msg/coal-capstone/nXhRb1fA7lg/sl8FNVCJAQAJ), i will also want to use **lazy evaluation** to avoid intermediate data. it may make more sense for me to move these methods into the mining id class if they don't end up being useful for environmental correlation.
* **mining ID** is my top priority right now, soft-blocked awaiting mineral classification data. i think the algorithm is sound, but there are a number of **subtleties** mentioned both in the code and that have occurred to me since. one of these is that we can't assume that pixels in two images have the same dimensions, so it may be necessary to scale them if the difference is large. another issue is the **machine-learning approach**: i still think neural nets are a good approach here, which brings us back to library selection. spectral's classifier is designed for multiband pixels, and although regions can be trivially rearranged to simulate these, it might make sense to choose a different library. **edit:** in fact, the maintainer of spectral [recommended](https://github.com/capstone-coal/pycoal/pull/37#issuecomment-285372637) we use a different NN implementation.

i still have next week, the week after, and spring break week after that before the start of next term so although i feel behind, i think there's a comfortable margin for dedicating a lot of time to the remaining issues. in addition, our TA has told us to expect to have a month or so next term that we can use for testing and revising. we can review the timeline in our meeting friday.

**Comment by ghost** [comment]
this also need tests.

**Pull request merged**
